File: src/0027.remove-element/main.c
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be
// found in the LICENSE file.

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

int removeElement(int* nums, int numsSize, int val) {
  for (int i = 0; i < numsSize; /* empty */) {
    if (nums[i] == val) {
      // Implements swap_remove()
      nums[i] = nums[numsSize - 1];
      numsSize -= 1;
    } else {
      i += 1;
    }
  }
  return numsSize;
}

void checkSolution(int* nums, int numsSize, int val, int expected_k) {
  int k = removeElement(nums, numsSize, val);
  for (int i = 0; i < k; ++i) {
    printf("%d, ", nums[i]);
  }
  printf("\n");
  assert(k == expected_k);
}

int main() {
  int arr1[] = {0, 1, 2, 2, 3, 0, 4, 2};
  checkSolution(arr1, sizeof(arr1) / sizeof(arr1[0]), 2, 5);

  int arr2[] = {3, 2, 2, 3};
  checkSolution(arr2, sizeof(arr2) / sizeof(arr2[0]), 3, 2);

  return 0;
}

File: src/0027.remove-element/src/main.rs
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be
// found in the LICENSE file.

pub fn remove_element(nums: &mut Vec<i32>, val: i32) -> i32 {
    let mut i = 0;
    while i < nums.len() {
        if nums[i] == val {
            nums.swap_remove(i);
        } else {
            i += 1;
        }
    }
    nums.len() as i32
}

fn main() {
    let mut nums = vec![3, 2, 2, 3];
    let val = 3;
    let len = remove_element(&mut nums, val);
    assert_eq!(len, 2);
    assert_eq!(nums, vec![2, 2]);

    let mut nums = vec![0, 1, 2, 2, 3, 0, 4, 2];
    let val = 2;
    let len = remove_element(&mut nums, val);
    assert_eq!(len, 5);
    assert_eq!(nums, vec![0, 1, 4, 0, 3]);
}
